// react-dnd-card/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ItemTypes = require('./ItemTypes');

var _ItemTypes2 = _interopRequireDefault(_ItemTypes);

var _reactDnd = require('react-dnd');

var _lodash = require('lodash.flow');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function isNullOrUndefined(o) {
  return o == null;
}

var cardSource = {
  beginDrag: function beginDrag(props) {
    return {
      id: props.id,
      index: props.index,
      originalIndex: props.index
    };
  },
  endDrag: function endDrag(props, monitor) {
    if (props.noDropOutside) {
      var _monitor$getItem = monitor.getItem();

      var id = _monitor$getItem.id;
      var index = _monitor$getItem.index;
      var originalIndex = _monitor$getItem.originalIndex;

      var didDrop = monitor.didDrop();

      if (!didDrop) {
        props.moveCard(isNullOrUndefined(id) ? index : id, originalIndex);
      }
    }

    if (props.endDrag) {
      props.endDrag();
    }
  }
};

var cardTarget = {
  hover: function hover(props, monitor) {
    var _monitor$getItem2 = monitor.getItem();

    var dragId = _monitor$getItem2.id;
    var dragIndex = _monitor$getItem2.index;
    var hoverId = props.id;
    var hoverIndex = props.index;


    if (!isNullOrUndefined(dragId)) {
      // use id
      if (dragId !== hoverId) {
        props.moveCard(dragId, hoverIndex);
      }
    } else {
      // use index
      if (dragIndex !== hoverIndex) {
        props.moveCard(dragIndex, hoverIndex);

        // Note: we're mutating the monitor item here!
        // Generally it's better to avoid mutations,
        // but it's good here for the sake of performance
        // to avoid expensive index searches.
        monitor.getItem().index = hoverIndex;
      }
    }
  }
};

var propTypes = {
  index: _react.PropTypes.number.isRequired,
  source: _react.PropTypes.any.isRequired,
  createItem: _react.PropTypes.func.isRequired,
  moveCard: _react.PropTypes.func.isRequired,
  endDrag: _react.PropTypes.func,
  isDragging: _react.PropTypes.bool.isRequired,
  connectDragSource: _react.PropTypes.func.isRequired,
  connectDropTarget: _react.PropTypes.func.isRequired,
  noDropOutside: _react.PropTypes.bool
};

var DndCard = function (_Component) {
  _inherits(DndCard, _Component);

  function DndCard() {
    _classCallCheck(this, DndCard);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(DndCard).apply(this, arguments));
  }

  _createClass(DndCard, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var id = _props.id;
      var index = _props.index;
      var source = _props.source;
      var createItem = _props.createItem;
      var noDropOutside = _props.noDropOutside;
      var // remove from restProps
      moveCard = _props.moveCard;
      var // remove from restProps
      endDrag = _props.endDrag;
      var // remove from restProps

      isDragging = _props.isDragging;
      var connectDragSource = _props.connectDragSource;
      var connectDropTarget = _props.connectDropTarget;

      var restProps = _objectWithoutProperties(_props, ['id', 'index', 'source', 'createItem', 'noDropOutside', 'moveCard', 'endDrag', 'isDragging', 'connectDragSource', 'connectDropTarget']);

      if (id === null) {
        console.warn('Warning: `id` is null. Set to undefined to get better performance.');
      }

      var item = createItem(source, isDragging, index);
      if (typeof item === 'undefined') {
        console.warn('Warning: `createItem` returns undefined. It should return a React element or null.');
      }

      var finalProps = Object.keys(restProps).reduce(function (result, k) {
        var prop = restProps[k];
        result[k] = typeof prop === 'function' ? prop(isDragging) : prop;
        return result;
      }, {});

      return connectDragSource(connectDropTarget(_react2.default.createElement(
        'div',
        finalProps,
        item
      )));
    }
  }]);

  return DndCard;
}(_react.Component);

DndCard.propTypes = propTypes;

exports.default = (0, _lodash2.default)((0, _reactDnd.DropTarget)(_ItemTypes2.default.DND_CARD, cardTarget, function (connect) {
  return {
    connectDropTarget: connect.dropTarget()
  };
}), (0, _reactDnd.DragSource)(_ItemTypes2.default.DND_CARD, cardSource, function (connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging()
  };
}))(DndCard);
